src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryMutation.cs

```diff
@@ -229,6 +232,111 @@ public async Task<FormulaValue> InvokeAsync(FormulaType irContext, FormulaValue[
+        {
+            return await new CollectProcess().Process(irContext, args, cancellationToken).ConfigureAwait(false);
+        }
+    }
+
+    internal class CollectProcess
+    {
+        internal async Task<FormulaValue> Process(FormulaType irContext, FormulaValue[] args, CancellationToken cancellationToken)
+        {
+            FormulaValue arg0;
+            var argc = args.Length;
+
+            // Need to check if the Lazy first argument has been evaluated since it may have already been
+            // evaluated in the ClearCollect case.
+            if (args[0] is LambdaFormulaValue arg0lazy)
+            {
+                arg0 = await arg0lazy.EvalAsync().ConfigureAwait(false);
+            }
+            else
+            {
+                arg0 = args[0];
+            }
+
+            if (arg0 is BlankValue)
+            {
+                return arg0;
+            }
+
+            if (arg0 is ErrorValue)
+            {
+                return arg0;
+            }
+
+            if (arg0 is not TableValue)
+            {
+                return CommonErrors.RuntimeTypeMismatch(IRContext.NotInSource(arg0.Type));
+            }
+
+            var tableValue = arg0 as TableValue;
+
+            List<DValue<RecordValue>> resultRows = new List<DValue<RecordValue>>();
+
+            for (int i = 1; i < argc; i++)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var arg = args[i];
+
+                if (arg is TableValue argTableValue)
+                {
+                    foreach (DValue<RecordValue> row in argTableValue.Rows)
+                    {
+                        if (row.IsBlank)
+                        {
+                            continue;
+                        }
+                        else if (row.IsError)
+                        {
+                            return row.Error;
+                        }
+                        else
+                        {
+                            var recordValueCopy = (RecordValue)row.ToFormulaValue().MaybeShallowCopy();
+                            resultRows.Add(await tableValue.AppendAsync(recordValueCopy, cancellationToken).ConfigureAwait(false));
+                        }
+                    }
+                }
+                else if (arg is RecordValue)
+                {
+                    var recordValueCopy = CompileTimeTypeWrapperRecordValue.AdjustType(tableValue.Type.ToRecord(), (RecordValue)arg.MaybeShallowCopy());
+                    resultRows.Add(await tableValue.AppendAsync(recordValueCopy, cancellationToken).ConfigureAwait(false));
+                }
+                else if (arg is ErrorValue)
+                {
+                    return arg;
+                }
+            }
+
+            if (resultRows.Count == 0)
+            {
+                return FormulaValue.NewBlank(arg0.Type);
```

Comment:
>arg0.Type [](http://example.com/codeflow?start=45&length=9)

This may not be the correct type for all cases. Assuming that `coll` is a collection with schema `*[a:s]`, then those should be the result for the following expressions:

- `Collect(coll, If(1<0, { a:"hello" }))`
  - `![a:s]` - single record, record result
- `Collect(coll, If(1<0, [{ a:"hello" }]))`
  - `*[a:s]` - table or multiple records, table result